aptos-move/framework/src/natives/aptos_lib.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::{
+    gas_schedule::GasCost,
+    language_storage::{StructTag, TypeTag},
+};
+use move_vm_runtime::native_functions::NativeContext;
+use move_vm_types::{
+    loaded_data::runtime_types::Type, natives::function::NativeResult, values::Value,
+};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, fmt::Write};
+
+/// Return Token types ModuleAddress, ModuleName and StructName
+pub fn type_of(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(ty_args.len() == 1);
+    debug_assert!(arguments.is_empty());
+
+    let cost = GasCost::new(super::cost::APTOS_LIB_TYPE_OF, 1).total();
+
+    let type_tag = context.type_to_type_tag(&ty_args[0])?;
+    if let TypeTag::Struct(struct_tag) = type_tag {
+        Ok(NativeResult::ok(
+            cost,
+            type_of_internal(&struct_tag).expect("type_of should never fail."),
+        ))
+    } else {
+        Ok(NativeResult::err(
+            cost,
+            super::status::NFE_EXPECTED_STRUCT_TYPE_TAG,
+        ))
+    }
+}
+
+fn type_of_internal(struct_tag: &StructTag) -> Result<SmallVec<[Value; 1]>, std::fmt::Error> {
+    let mut name = struct_tag.name.to_string();
+    if let Some(first_ty) = struct_tag.type_params.first() {
+        write!(name, "<")?;
+        write!(name, "{}", first_ty)?;
+        for ty in struct_tag.type_params.iter().skip(1) {
+            write!(name, ", {}", ty)?;
+        }
+        write!(name, ">")?;
+    }
+
+    Ok(smallvec![
+        Value::address(struct_tag.address),
+        Value::vector_u8(struct_tag.module.as_bytes().to_vec()),
+        Value::vector_u8(name.as_bytes().to_vec()),
+    ])
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use move_core_types::{account_address::AccountAddress, identifier::Identifier};
+
+    #[test]
+    fn test_type_of_internal() {
```

Comment:
I think it deserves a unit test in Move?

Reply:
ha, that's why the validator build probably failed...